h/security/permits/engine.py

```diff
@@ -0,0 +1,69 @@
+from enum import Enum
+
+
+def predicate(requires=None):
+    def decorator(function):
+        function.requires = requires or []
+        return function
+
+    return decorator
+
+
+class PredicateBasedPermissions:
+    def __init__(self, mapping):
+        self.mapping = self._expand_clauses(mapping)
+
+    def permits(self, identity, context, permission):
+        if clauses := self.mapping.get(permission):
+            statement_cache = {}
+
+            for clause in clauses:
+                if self._clause_true(clause, statement_cache, identity, context):
+                    return True
+
+            return False
+
+        return False
+
+    def _clause_true(self, clause, statement_cache, identity, context):
+        for predicate in clause:
+            value = statement_cache.get(predicate)
+            if value is None:
+                if isinstance(predicate, Enum):
+                    # We are going to assume this predicate is a permission
+                    value = self.permits(identity, context, predicate)
```

Comment:
Might be swell to have a `PermissionEnum` class so we can tell the difference between our permissions and any old `Enum`

Reply:
I think you could do this the other way round: try calling `predicate(identity, context)` and if you get a `TypeError` (because `predicate` isn't callable) then try `self.permits(identity, context, predicate)`. That would work with any type of permission, e.g. neither strings nor enums are callable.

It might be kind of nice for `PredicateBasedPermissions` to be ignorant of the fact that we're using enums for permissions?

Reply:
Maybe, but also we have a fixed Enum for permissions partially so they can't be confused with anything else. Seems like a good property to make use of?

Anything else in this location might cause some pretty confusing outcomes.